docs: rewrite the data model and schema design docs (#1590)
GreptimeTeam/docs: docs/user-guide/administration/design-table.md

- **Low Cardinality**: Low cardinality columns usually have constant values.
  The total number of unique values usually no more than 10 thousand.
  For example, `namespace`, `cluster`, `http_method` are usually low cardinality.
- **High Cardinality**: High cardinality columns contain a large number of unique values.
  For example, `trace_id`, `span_id`, `user_id`, `uri`, `ip`, `uuid`, `request_id`, table auto increment id are usually high cardinality.

## Column Types and Selection

In GreptimeDB, columns are categorized into three semantic types: `Tag`, `Field`, and `Timestamp`.
The timestamp usually represents the time of data sampling or the occurrence time of logs/events.
GreptimeDB uses the `TIME INDEX` constraint to identify the `Timestamp` column.
So the `Timestamp` column is also called the `TIME INDEX` column.
If you have multiple columns with timestamp data type, you can only define one as `TIME INDEX` and others as `Field` columns.

In GreptimeDB, tag columns are optional. The main purposes of Tag columns include:

1. Defining the ordering of data in storage.
   GreptimeDB reuses the `PRIMARY KEY` constraint to define tag columns and the ordering of tags.

>    Unlike traditional databases, in GreptimeDB, primary key and time series are identical.

Reviewer: ```suggestion
   Unlike traditional databases, in GreptimeDB, the primary key and time series in typical TSDB are identical.
```

Reviewer 2: > primary key and time series are identical

We can improve this by time series is defined by primary key

>    Tables in GreptimeDB sort rows by (primary key, timestamp).

Reviewer: ```suggestion
   Tables in GreptimeDB sort rows by in the order of `(primary key, timestamp)`.
```

   This improves the locality of data with the same tags.
   If there are no tag columns, GreptimeDB sorts rows by timestamp.
2. Identifying a unique time-series.

>    When the table is not append-only, GreptimeDB can deduplicate rows by timestamp under the same time-series (primary key).

Author: Append mode is intended not to dedup, some data like logs can have duplicate keys.

>    When the table is not append-only, GreptimeDB can deduplicate rows by timestamp under the same time-series (primary key).

Reviewer 2: I think we should at least keep a semantic for deduplicate entries. For performance consideration, we can do best-effort dedup, for example dedup on read or on compaction. Otherwise we will be in a position that in theory we cannot deal with logs generated from upstream retries.

>    When the table is not append-only, GreptimeDB can deduplicate rows by timestamp under the same time-series (primary key).

Author: In most analytical databases, the primary key is not unique.

> 3. Smoothing migration from other TSDBs that use tags or labels.

Reviewer 2: We need to mention the impact of the order of tags in primary key

Author: The `When to use primary key` section explains this and has an example.

### Choosing tag columns

Usually you don't need tag columns since ordering data by timestamp is sufficient for most use cases.
If you need deduplication, or your queries can benefit from the ordering, you can define tag columns.

> Setting tags may also reduce disk space usage as it improves the locality of data.

Reviewer 2: I feel this is due to the limitation caused cardinality effect on memtable. This should not be, at least the final, definition of tags. 

The selection of tags determines:

1. how we can dedup data
2. how we can sort data to narrow down the scan range of user's most typical queries
3. how we can do effective partition, for small query there should be less partitions envolved.
4. sorted data brings better compression as mentioned

Author: Removing the limitation may be the final goal, but this document is a guide for the current version.

However, a high cardinality primary key may still cause issues in some time-series workloads, like fetching last rows, processing in per-series manner. This may be related to the schema design.

Author: Not only memtable, our batch struct is also per series.
https://github.com/GreptimeTeam/greptimedb/blob/c77ce958a3b936e1065abe51fa10832b0ea2ac99/src/mito2/src/read.rs#L63-L74

In order to improve sort and deduplication speed under time-series workloads, GreptimeDB buffers and processes rows by time-series internally.
So it doesn't need to compare the primary key for each row repeatedly.
This can be a problem if the tag column has high cardinality:

1. Performance degradation since the database can't batch rows efficiently.
2. It may increase memory and CPU usage as the database has to maintain the metadata for each time-series.
3. Deduplication may be too expensive.

Recommendations:

- Low cardinality columns that occur in `WHERE`/`GROUP BY`/`ORDER BY` frequently.
  These columns usually remain constant.
  For example, `namespace`, `cluster`, or an AWS `region`.
- No need to set all low cardinality columns as tags since this may impact the performance of ingestion and querying.
- Typically use short strings for tags, avoiding `FLOAT`, `DOUBLE`, `TIMESTAMP`.
- Never set high cardinality columns as tags if they change frequently.
  For example, `trace_id`, `span_id`, `user_id` must not be used as tags.

>   GreptimeDB works well if you set them as fields instead of tags.

Reviewer: Note: Creating indexes on fields and tags can be done at any time.

Reviewer 2: Some content for reference: https://clickhouse.com/docs/use-cases/observability/schema-design#choosing-a-primary-ordering-key

Author: Currently, this is an implementation limitation. Our primary key is a time-series and is similar to a stream in VL.
https://docs.victoriametrics.com/victorialogs/keyconcepts/#how-to-determine-which-fields-must-be-associated-with-log-streams

In the future, we can support flexible primary key, but some optimization is only possible and available when the cardinality is low,

For `trace_id+spanid`, it is actually not a time-series. We may need to consider this in the future.

## Primary Key and Index

Bad primary key or index may significantly degrade performance.
Generally you can create an append-only table without a primary key for most workloads.
It can also serve as a baseline.

```sql
CREATE TABLE http_logs (
  access_time TIMESTAMP TIME INDEX,
  application STRING,
  remote_addr STRING,
  http_status STRING,
  http_method STRING,
  http_refer STRING,
  user_agent STRING,
  request_id STRING,
  request STRING,
) with ('append_mode'='true');
```

The `http_logs` table is an example for storing HTTP server logs.

- The `'append_mode'='true'` option creates the table as an append-only table.
  This ensures a log doesn't override another one with the same timestamp.
- The table sorts logs by time so it is efficient to search logs by time.

### When to use primary key

> In GreptimeDB, the primary key is identical to the time-series.

Reviewer: ```suggestion
In GreptimeDB, the primary key is identical to the time-series in typical TSDB.
```

It defines tag columns and the data ordering.
You can use primary key when there are suitable low cardinality columns and one of the following conditions is met:

- Most queries can benefit from the ordering.
- You need to deduplicate (including delete) rows by the primary key and time index.

For example, if you always only query logs of a specific application, you may set the `application` column as primary key (tag).

```sql
SELECT message FROM http_logs WHERE application = 'greptimedb' AND access_time > now() - '5 minute'::INTERVAL;
```

The number of applications is usually limited. Table `http_logs_v2` uses `application` as a tag.
It sorts logs by application so querying logs under the same application is faster.

```sql
CREATE TABLE http_logs_v2 (
  access_time TIMESTAMP TIME INDEX,
  application STRING,
  remote_addr STRING,
  http_status STRING,
  http_method STRING,
  http_refer STRING,
  user_agent STRING,
  request_id STRING,
  request STRING,
  PRIMARY KEY(application),
) with ('append_mode'='true');
```

> The unique values of the primary key in the table can significantly affect performance.

Reviewer: ```suggestion
The unique values (AKA cardinality) of the primary key in the table can significantly affect performance.
```

Currently, the recommended number of values for the primary key is no more than 100 thousand.
So you must not use high cardinality column as the primary key or put too many columns in the primary key.

Besides primary key, you can also use index to accelerate specific queries on demand.

### Inverted Index

GreptimeDB supports inverted index that may speed up filtering low cardinality columns.
When creating a table, you can specify the [inverted index](/contributor-guide/datanode/data-persistence-indexing.md#inverted-index) columns using the `INVERTED INDEX` clause.
For example, `http_logs_v2` adds an inverted index for the `http_method` column.

```sql
CREATE TABLE http_logs_v2 (
  access_time TIMESTAMP TIME INDEX,
  application STRING,
  remote_addr STRING,
  http_status STRING,
  http_method STRING INVERTED INDEX,
  http_refer STRING,
  user_agent STRING,
  request_id STRING,
  request STRING,
  PRIMARY KEY(application),
) with ('append_mode'='true');
```

The following query can use the inverted index on the `http_method` column.

> ```sql

Reviewer: Do we need to explain which type of queries can benefit from indexes? Only the exact equal or prefix matching or range queries like that.

SELECT message FROM http_logs WHERE application = 'greptimedb' AND http_method = `GET` AND access_time > now() - '5 minute'::INTERVAL;
```

### Skipping Index

For high cardinality columns like `trace_id`, `request_id`, using a [skipping index](/user-guide/manage-data/data-index.md#skipping-index) is more appropriate.

> This method has lower storage overhead and resource usage, particularly in terms of memory consumption.

Reviewer: ```suggestion
This method has lower storage overhead and resource usage, particularly in terms of memory and disk consumption.
```

### When to Partition and Determining the Number of Partitions

A table can utilize all the resources in the machine, especially during query.
Partitioning a table may not always improve the performance:

- A distributed query plan isn't always as efficient as a local query plan.
- Distributed query may introduce additional data transmission across the network.

There is no need to partition a table unless a single machine isn't enough to serve the table.
For example:

- There is not enough local disk space to store the data or to cache the data when using object stores.
- You need more CPU cores to improve the query performance or more memory for costly queries.
- The disk throughput becomes the bottleneck.
- The ingestion rate is larger than the throughput of a single node.

> GreptimeDB releases a [benchmark report](https://github.com/GreptimeTeam/greptimedb/tree/main/docs/benchmarks/tsbs) with each major version update,

Reviewer 3: ```suggestion
GreptimeDB releases a [benchmark report](https://github.com/GreptimeTeam/greptimedb/tree/VAR::greptimedbVersion/docs/benchmarks/tsbs) with each major version update,
```
